app/utils/ContextErrorMaping.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2019 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package utils
+
+import play.api.data.{FormError, Mapping}
+import play.api.data.validation.Constraint
+
+case class ContextErrorMapping[T](origin: Mapping[T]) extends Mapping[T] {
```

Comment:
I like the idea of wrapping the `Mapping` to provide extra e.g multi-field validation however I think it needs to be *intuitive* (or it wont be used) and I'm not convinced this is particularly.

I'd prefer if we could design the signature of what we want to do first e.g.

```
val mapping: Mapping[T] = 
Forms.mapping(
    "field1" -> text().verifying("field1.error", isEmpty or isValidField),
    "field2" -> text().verifying("field2.error", isEmpty or isValidField)
)(SomeClass.apply)(SomeClass.unapply)
    .verifying(Seq("field1.error", "field2.error"), someClass => isNonEmpty(someClass.field1) or isNonEmpty(someClass.field2))
```

Then make it work.

e.g. with a wrapper class e.g.
```
implicit class ContextErrorMapping[T](origin: Mapping[T]) extends Mapping[T] {
      def verifying(fieldErrors: Seq[String], condition: T => Boolean) = {
            // Some code to add the errors
      }
}
```

That way it stays intuitive

Reply:
But here we won't produce error at two field at one. Put in point - API based on convention and there is no straight obvious way to use it. 

I will migrate it to solution by conditional mapping